packages/mdc-textfield/input/foundation.js

```diff
@@ -0,0 +1,181 @@
+    /** @private {function(): undefined} */
+    this.inputInputHandler_ = () => this.autoCompleteFocus_();
+    /** @private {function(): undefined} */
+    this.inputPressedHandler_ = () => this.notifyPressedAction_();
+  }
+
+  init() {
+    this.adapter_.registerEventHandler('focus', this.inputFocusHandler_);
+    this.adapter_.registerEventHandler('blur', this.inputBlurHandler_);
+    this.adapter_.registerEventHandler('input', this.inputInputHandler_);
+    ['mousedown', 'touchstart'].forEach((evtType) => {
+      this.adapter_.registerEventHandler(evtType, this.inputPressedHandler_);
+    });
+  }
+
+  destroy() {
+    this.adapter_.deregisterEventHandler('focus', this.inputFocusHandler_);
+    this.adapter_.deregisterEventHandler('blur', this.inputBlurHandler_);
+    this.adapter_.deregisterEventHandler('input', this.inputInputHandler_);
+    ['mousedown', 'touchstart'].forEach((evtType) => {
+      this.adapter_.deregisterEventHandler(evtType, this.inputPressedHandler_);
+    });
+  }
+
+  /**
+   * Activates the input's focus state.
+   * @private
+   */
+  activateFocus_() {
+    this.adapter_.notifyFocusAction();
+  }
+
+  /**
+   * Activates the input's focus state in cases when the input value
+   * changes without user input (e.g. programatically).
+   * @private
+   */
+  autoCompleteFocus_() {
+    if (!this.receivedUserInput_) {
+      this.activateFocus_();
+    }
+  }
+
+  /**
+   * Deactives the input's focus state.
+   * @private
+   */
+  deactivateFocus_() {
+    const input = this.getNativeInput_();
+
+    if (!input.value && !this.isBadInput()) {
+      this.receivedUserInput_ = false;
+    }
+
+    this.adapter_.notifyBlurAction();
+  }
+
+  notifyPressedAction_() {
+    this.adapter_.notifyPressedAction();
+  }
+
+  /**
+   * @return {boolean} True if the input fails validity checks.
+   */
+  isBadInput() {
+    const input = this.getNativeInput_();
+    return input.validity ? input.validity.badInput : input.badInput;
+  }
+
+  /**
+   * @return {string} Returns the value in the input.
+   */
+  getValue() {
+    return this.getNativeInput_().value;
+  }
+
+  /**
+   * @return {boolean} Returns the result of the checkValidity() call in the native input.
+   */
+  checkValidity() {
+    return this.getNativeInput_().checkValidity();
+  }
+
+  /**
+   * @return {boolean} True if the input is disabled.
+   */
+  isDisabled() {
+    return this.getNativeInput_().disabled;
+  }
+
+  /**
+   * @param {boolean} disabled Sets the input disabled or enabled.
+   */
+  setDisabled(disabled) {
+    this.getNativeInput_().disabled = disabled;
+  }
+
+  /**
+   * Handles text field interaction.
+   */
+  handleTextFieldInteraction() {
+    this.receivedUserInput_ = true;
+  }
+
+  /**
+   * @return {!Element|!NativeInputType} The native text input from the
+   * host environment, or a dummy if none exists.
+   * @private
+   */
+  getNativeInput_() {
+    return this.adapter_.getNativeInput() ||
```

Comment:
I realize this existed prior to the split, but...is there ever not an input element??

Reply:
Its annotated in the `MDCTextFieldInputAdapter` as `{?Element|?NativeInputType}`. So technically it is optional. But again, lets not architecture in this PR. This PR is just refactoring. If it illuminates some problems with the old MDCTextFieldFoundation code...great! But file issues and we will resolve in another PR.